File: core/IsographicQuaternion.cs
using System;

namespace VAN_Engine.Core
{
    /// <summary>
    /// Isographic Quaternion with coupled dimensions (Sound-Shape, Sound-Number, Shape-Time, Number-Time)
    /// ISO_009: Quadruple Mapping
    /// </summary>
    public struct IsographicQuaternion
    {
        public double W { get; set; } // Sound-Shape coupling
        public double X { get; set; } // Sound-Number coupling
        public double Y { get; set; } // Shape-Time coupling
        public double Z { get; set; } // Number-Time coupling

        public IsographicQuaternion(double w, double x, double y, double z)
        {
            W = w;
            X = x;
            Y = y;
            Z = z;
        }

        // Projection methods for specific queries
        public double GetSoundProjection() => Math.Sqrt(W * W + X * X);
        public double GetShapeProjection() => Math.Sqrt(W * W + Y * Y);
        public double GetNumberProjection() => Math.Sqrt(X * X + Z * Z);
        public double GetTimeProjection() => Math.Sqrt(Y * Y + Z * Z);

        // Global magnitude (must preserve across round-trips)
        public double Magnitude => Math.Sqrt(W * W + X * X + Y * Y + Z * Z);

        // Normalization
        public IsographicQuaternion Normalize()
        {
            double mag = Magnitude;
            if (mag < 1e-10) return this;
            return new IsographicQuaternion(W / mag, X / mag, Y / mag, Z / mag);
        }

        // Quaternion multiplication (for transit modifiers)
        public static IsographicQuaternion operator *(IsographicQuaternion a, IsographicQuaternion b)
        {
            return new IsographicQuaternion(
                a.W * b.W - a.X * b.X - a.Y * b.Y - a.Z * b.Z,
                a.W * b.X + a.X * b.W + a.Y * b.Z - a.Z * b.Y,
                a.W * b.Y - a.X * b.Z + a.Y * b.W + a.Z * b.X,
                a.W * b.Z + a.X * b.Y - a.Y * b.X + a.Z * b.W
            );
        }

        // Equality with tolerance
        public bool ApproxEquals(IsographicQuaternion other, double tolerance = 1e-6)
        {
            return Math.Abs(W - other.W) < tolerance &&
                   Math.Abs(X - other.X) < tolerance &&
                   Math.Abs(Y - other.Y) < tolerance &&
                   Math.Abs(Z - other.Z) < tolerance;
        }

        public override string ToString() => $"Q({W:F3}, {X:F3}, {Y:F3}, {Z:F3})";
    }
}

File: engine/core/IsographicQuaternion.py
"""
Isographic Quaternion with coupled dimensions (Sound-Shape, Sound-Number, Shape-Time, Number-Time)
ISO_009: Quadruple Mapping
Windows UTF-8 safe
"""

import sys
import io
import math

if sys.platform == 'win32':
    if hasattr(sys.stdout, 'buffer') and sys.stdout.buffer and not isinstance(sys.stdout, io.TextIOWrapper):
        sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
    if hasattr(sys.stderr, 'buffer') and sys.stderr.buffer and not isinstance(sys.stderr, io.TextIOWrapper):
        sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8')


class IsographicQuaternion:
    """4D quaternion with coupled dimensions for isographic indexing"""

    def __init__(self, w: float, x: float, y: float, z: float):
        self.w = w  # Sound-Shape coupling
        self.x = x  # Sound-Number coupling
        self.y = y  # Shape-Time coupling
        self.z = z  # Number-Time coupling

    def get_sound_projection(self) -> float:
        return math.sqrt(self.w * self.w + self.x * self.x)

    def get_shape_projection(self) -> float:
        return math.sqrt(self.w * self.w + self.y * self.y)

    def get_number_projection(self) -> float:
        return math.sqrt(self.x * self.x + self.z * self.z)

    def get_time_projection(self) -> float:
        return math.sqrt(self.y * self.y + self.z * self.z)

    @property
    def magnitude(self) -> float:
        return math.sqrt(self.w * self.w + self.x * self.x + self.y * self.y + self.z * self.z)

    def normalize(self) -> 'IsographicQuaternion':
        mag = self.magnitude
        if mag < 1e-10:
            return self
        return IsographicQuaternion(self.w / mag, self.x / mag, self.y / mag, self.z / mag)

    def __mul__(self, other: 'IsographicQuaternion') -> 'IsographicQuaternion':
        return IsographicQuaternion(
            self.w * other.w - self.x * other.x - self.y * other.y - self.z * other.z,
            self.w * other.x + self.x * other.w + self.y * other.z - self.z * other.y,
            self.w * other.y - self.x * other.z + self.y * other.w + self.z * other.x,
            self.w * other.z + self.x * other.y - self.y * other.x + self.z * other.w
        )

    def approx_equal(self, other: 'IsographicQuaternion', tolerance: float = 1e-6) -> bool:
        return (abs(self.w - other.w) < tolerance and
                abs(self.x - other.x) < tolerance and
                abs(self.y - other.y) < tolerance and
                abs(self.z - other.z) < tolerance)

    def __repr__(self) -> str:
        return f"IsographicQuaternion(w={self.w:.3f}, x={self.x:.3f}, y={self.y:.3f}, z={self.z:.3f})"
